In [1]:
import pandas as pd
import numpy as np
from zipfile import ZipFile 
from functools import reduce 
import matplotlib.pyplot as plt
from scipy import stats 
import os 

In [2]:
# this is a zip file containing multiple csvs'
path = r"C:\Users\ncc\Downloads\archive(1).zip"
extracted_file = "data_files"

In [3]:
with ZipFile(path, "r") as f:
    f.extractall("data_files")

In [4]:
data_frames = {}

for file in os.listdir(extracted_file):
    if file.endswith("csv"):
        file_path = os.path.join(extracted_file, file)
        data_frames[file] = pd.read_csv(file_path)

In [5]:
for name, df in data_frames.items():
    print(f"{name}")

olist_customers_dataset.csv
olist_geolocation_dataset.csv
olist_orders_dataset.csv
olist_order_items_dataset.csv
olist_order_payments_dataset.csv
olist_order_reviews_dataset.csv
olist_products_dataset.csv
olist_sellers_dataset.csv
product_category_name_translation.csv


In [6]:
customers = pd.read_csv(r"data_files/olist_customers_dataset.csv")
customers.head()
print(customers.shape)
print(customers.columns)

(99441, 5)
Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='object')


In [7]:
geolocation = pd.read_csv(r"data_files\olist_geolocation_dataset.csv")
geolocation.head()
print(geolocation.shape)
print(geolocation.columns)

(1000163, 5)
Index(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state'],
      dtype='object')


In [8]:
order_items = pd.read_csv(r"data_files\olist_order_items_dataset.csv")
order_items.head()
print(order_items.shape)
print(order_items.columns)

(112650, 7)
Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='object')


In [9]:
order_payment = pd.read_csv(r"data_files\olist_order_payments_dataset.csv")
order_payment.head()
print(order_payment.columns)
print(order_payment.shape)

Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='object')
(103886, 5)


In [10]:
order_reviews = pd.read_csv(r"data_files\olist_order_reviews_dataset.csv")
order_reviews.head()
print(order_reviews.columns)
print(order_reviews.shape)

Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='object')
(99224, 7)


In [11]:
order_data = pd.read_csv(r"data_files\olist_orders_dataset.csv")
order_data.head()
print(order_data.columns)
print(order_data.shape)

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')
(99441, 8)


In [12]:
product = pd.read_csv(r"data_files\olist_products_dataset.csv")
product.head()
print(product.columns)
print(product.shape)

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='object')
(32951, 9)


In [14]:
sellers = pd.read_csv(r"data_files\olist_sellers_dataset.csv")
sellers.head()
print(sellers.columns)
print(sellers.shape)

Index(['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state'], dtype='object')
(3095, 4)


In [15]:
product_category = pd.read_csv(r"data_files\product_category_name_translation.csv")
product_category.head()
print(product_category.columns)
print(product_category.shape)

Index(['product_category_name', 'product_category_name_english'], dtype='object')
(71, 2)


In [16]:
for name, df in data_frames.items():
    print(f"{name}: {list(df.columns)}")

olist_customers_dataset.csv: ['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']
olist_geolocation_dataset.csv: ['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']
olist_orders_dataset.csv: ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']
olist_order_items_dataset.csv: ['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']
olist_order_payments_dataset.csv: ['order_id', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value']
olist_order_reviews_dataset.csv: ['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']
olist_products_dataset.csv: ['product_id', 'product_category_name', 'pro

In [17]:
# get sets of columns for each file
columns_per_file = {name: set(df.columns) for name, df in data_frames.items()}

# intersection (columns that appear in every file)
common_cols = reduce(set.intersection, columns_per_file.values())
print("Common columns across all CSVs:", common_cols)

Common columns across all CSVs: set()


**MERGING DATASET AND WHY THEY SHOULD BE MERGED**
1.  **Customers** <--> **Orders**
-   **Key:** `customer_id`
-   **Why:** Orders are made by customers. You can atatch customer info (city, state, etc.) to each other.

Merge: `order_data` <--> `customers`

2.  **Orders** <--> **Order Items**
-   **Key:** `order_id`
-   **Why:** Each order has one or more items. This link orders to products and sellers.

Merge: `order_data` <--> `order_items`

3.  **Order Items** <--> **Products**
-   **Key:** `product_id`
-   **Why:** Each item corresponds to a product. You can bring in product details (category, dimensions, weight).

Merge: `order_items` <--> `product`

4.  **Products** <--> **Product Category Translation**
-   **Key:** `product_category_name`
-   **Why:** Product categories are in Portuguese, and this dataset provides English translations.

Merge: `product` <--> `product_category`

5.  **Order Items** <--> **Sellers**
-   **Key:** `seller_id`
-   **Why:** Each item is sold by a seller. This lets you connect sellers' locations to orders.

Merge: `order_items` <--> `sellers`

6.  **Orders** <--> **Payments**
-   **Key:** `order_id`
-   **Why:** Every order has a payment record (method, value, installments).

Merge: `order_data` <--> `order_payment`

7.  **Orders** <--> **Reviews**
-   **Key:** `order_id`
-   **Why:** Customers leave reviews for orders (scores, comments)

Merge: `order_data` <--> `order_reviews`

8.  **Customers** <--> **Geolocation**
-   **Key:** `zip_code_prefix` (`customer_zip_code_previx` vs `geolocation_zip_code_prefix`)
-   **Why:** To attach latitude/longitude and city/state to customers. <br>NB: Multiple geolocations can share the same ZIP previx --> requires aggregation or nearest match.

Merge: `customers` <--> `geolocation`

In [18]:
order_data.head(2)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00

In [19]:
#1. cutomers + orders
orders_customers = pd.merge(order_data, customers, on = "customer_id", how = "right")
orders_customers.head()

order_id                       customer_id  \
0  00e7ee1b050b8499577073aeb2a297a1  06b8999e2fba1a1fbc88172c00ba8bc7   
1  29150127e6685892b6eab3eec79f59c7  18955e83d337fd6b2def6b18a428ac77   
2  b2059ed67ce144a36e2aa97d2c9e9ad2  4e7b3e00288586ebd08712fdd0374a03   
3  951670f92359f4fe4a63112aa7306eba  b2b6027bc5c5109e529d4dc6358b12c3   
4  6b7d50bd145f6fc7f33cebabd7e49d0f  4f2d8ab171c80ec8364f7c12e35b23ad   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-05-16 15:05:35  2017-05-16 15:22:12   
1    delivered      2018-01-12 20:48:24  2018-01-12 20:58:32   
2    delivered      2018-05-19 16:07:45  2018-05-20 16:19:10   
3    delivered      2018-03-13 16:06:38  2018-03-13 17:29:19   
4    delivered      2018-07-29 09:51:30  2018-07-29 10:10:09   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-05-23 10:47:57           2017-05-25 10:35:35   
1          2018-01-15 17:14:59           2018-01-29 12:41:19   
2          2018-06-11 14:31:00           2018-06-14 17:58:51   
3          2018-03-27 23:22:42           2018-03-28 16:04:25   
4          2018-07-30 15:16:00           2018-08-09 20:55:48   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-06-05 00:00:00  861eff4711a542e4b93843c6dd7febb0   
1           2018-02-06 00:00:00  290c77bc529b7ac935b93aa66c333dc3   
2           2018-06-13 00:00:00  060e732b5b29e8181a18229c7b0b2b5e   
3           2018-04-10 00:00:00  259dac757896d24d7702b9acbbff3f3c   
4           2018-08-15 00:00:00  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [20]:
orders_customers.shape

(99441, 12)

In [21]:
#2. Orders + payments
orders_payments = pd.merge(orders_customers, order_payment, on = "order_id", how = "left")
orders_payments.head()

order_id                       customer_id  \
0  00e7ee1b050b8499577073aeb2a297a1  06b8999e2fba1a1fbc88172c00ba8bc7   
1  29150127e6685892b6eab3eec79f59c7  18955e83d337fd6b2def6b18a428ac77   
2  b2059ed67ce144a36e2aa97d2c9e9ad2  4e7b3e00288586ebd08712fdd0374a03   
3  951670f92359f4fe4a63112aa7306eba  b2b6027bc5c5109e529d4dc6358b12c3   
4  6b7d50bd145f6fc7f33cebabd7e49d0f  4f2d8ab171c80ec8364f7c12e35b23ad   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-05-16 15:05:35  2017-05-16 15:22:12   
1    delivered      2018-01-12 20:48:24  2018-01-12 20:58:32   
2    delivered      2018-05-19 16:07:45  2018-05-20 16:19:10   
3    delivered      2018-03-13 16:06:38  2018-03-13 17:29:19   
4    delivered      2018-07-29 09:51:30  2018-07-29 10:10:09   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-05-23 10:47:57           2017-05-25 10:35:35   
1          2018-01-15 17:14:59           2018-01-29 12:41:19   
2          2018-06-11 14:31:00           2018-06-14 17:58:51   
3          2018-03-27 23:22:42           2018-03-28 16:04:25   
4          2018-07-30 15:16:00           2018-08-09 20:55:48   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-06-05 00:00:00  861eff4711a542e4b93843c6dd7febb0   
1           2018-02-06 00:00:00  290c77bc529b7ac935b93aa66c333dc3   
2           2018-06-13 00:00:00  060e732b5b29e8181a18229c7b0b2b5e   
3           2018-04-10 00:00:00  259dac757896d24d7702b9acbbff3f3c   
4           2018-08-15 00:00:00  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  \
0                     14409                 franca             SP   
1                      9790  sao bernardo do campo             SP   
2                      1151              sao paulo             SP   
3                      8775        mogi das cruzes             SP   
4                     13056               campinas             SP   

   payment_sequential payment_type  payment_installments  payment_value  
0                 1.0  credit_card                   2.0         146.87  
1                 1.0  credit_card                   8.0         335.48  
2                 1.0  credit_card                   7.0         157.73  
3                 1.0  credit_card                   1.0         173.30  
4                 1.0  credit_card                   8.0         252.25

In [22]:
#3. Orders + Reviews
orders_reviews = pd.merge(orders_payments, order_reviews, on = "order_id", how = "left")
orders_reviews.head()

order_id                       customer_id  \
0  00e7ee1b050b8499577073aeb2a297a1  06b8999e2fba1a1fbc88172c00ba8bc7   
1  29150127e6685892b6eab3eec79f59c7  18955e83d337fd6b2def6b18a428ac77   
2  b2059ed67ce144a36e2aa97d2c9e9ad2  4e7b3e00288586ebd08712fdd0374a03   
3  951670f92359f4fe4a63112aa7306eba  b2b6027bc5c5109e529d4dc6358b12c3   
4  6b7d50bd145f6fc7f33cebabd7e49d0f  4f2d8ab171c80ec8364f7c12e35b23ad   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-05-16 15:05:35  2017-05-16 15:22:12   
1    delivered      2018-01-12 20:48:24  2018-01-12 20:58:32   
2    delivered      2018-05-19 16:07:45  2018-05-20 16:19:10   
3    delivered      2018-03-13 16:06:38  2018-03-13 17:29:19   
4    delivered      2018-07-29 09:51:30  2018-07-29 10:10:09   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-05-23 10:47:57           2017-05-25 10:35:35   
1          2018-01-15 17:14:59           2018-01-29 12:41:19   
2          2018-06-11 14:31:00           2018-06-14 17:58:51   
3          2018-03-27 23:22:42           2018-03-28 16:04:25   
4          2018-07-30 15:16:00           2018-08-09 20:55:48   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-06-05 00:00:00  861eff4711a542e4b93843c6dd7febb0   
1           2018-02-06 00:00:00  290c77bc529b7ac935b93aa66c333dc3   
2           2018-06-13 00:00:00  060e732b5b29e8181a18229c7b0b2b5e   
3           2018-04-10 00:00:00  259dac757896d24d7702b9acbbff3f3c   
4           2018-08-15 00:00:00  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix  ... payment_sequential payment_type  \
0                     14409  ...                1.0  credit_card   
1                      9790  ...                1.0  credit_card   
2                      1151  ...                1.0  credit_card   
3                      8775  ...                1.0  credit_card   
4                     13056  ...                1.0  credit_card   

   payment_installments payment_value                         review_id  \
0                   2.0        146.87  88b8b52d46df026a9d1ad2136a59b30b   
1                   8.0        335.48  02fc48a9efa3e3d0f1a8ea26507eeec3   
2                   7.0        157.73  5ad6695d76ee186dc473c42706984d87   
3                   1.0        173.30  059a801bb31f6aab2266e672cab87bc5   
4                   8.0        252.25  8490879d58d6c5d7773f2739a03f089a   

   review_score review_comment_title  \
0           4.0                  NaN   
1           5.0                  NaN   
2           5.0                  NaN   
3           5.0                  NaN   
4           5.0        a melhor nota   

                          review_comment_message review_creation_date  \
0                                            NaN  2017-05-26 00:00:00   
1                                            NaN  2018-01-30 00:00:00   
2                                            NaN  2018-06-15 00:00:00   
3                                            NaN  2018-03-29 00:00:00   
4  O baratheon è esxelente Amo adoro o baratheon  2018-08-10 00:00:00   

  review_answer_timestamp  
0     2017-05-30 22:34:40  
1     2018-02-10 22:43:29  
2     2018-06-15 12:10:59  
3     2018-04-02 18:36:47  
4     2018-08-17 01:59:52  

[5 rows x 22 columns]

In [23]:
#4. Orders + Order Items
orders_items = pd.merge(orders_reviews, order_items, on = "order_id", how = "left")
orders_items.head()

order_id                       customer_id  \
0  00e7ee1b050b8499577073aeb2a297a1  06b8999e2fba1a1fbc88172c00ba8bc7   
1  29150127e6685892b6eab3eec79f59c7  18955e83d337fd6b2def6b18a428ac77   
2  b2059ed67ce144a36e2aa97d2c9e9ad2  4e7b3e00288586ebd08712fdd0374a03   
3  951670f92359f4fe4a63112aa7306eba  b2b6027bc5c5109e529d4dc6358b12c3   
4  6b7d50bd145f6fc7f33cebabd7e49d0f  4f2d8ab171c80ec8364f7c12e35b23ad   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-05-16 15:05:35  2017-05-16 15:22:12   
1    delivered      2018-01-12 20:48:24  2018-01-12 20:58:32   
2    delivered      2018-05-19 16:07:45  2018-05-20 16:19:10   
3    delivered      2018-03-13 16:06:38  2018-03-13 17:29:19   
4    delivered      2018-07-29 09:51:30  2018-07-29 10:10:09   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-05-23 10:47:57           2017-05-25 10:35:35   
1          2018-01-15 17:14:59           2018-01-29 12:41:19   
2          2018-06-11 14:31:00           2018-06-14 17:58:51   
3          2018-03-27 23:22:42           2018-03-28 16:04:25   
4          2018-07-30 15:16:00           2018-08-09 20:55:48   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-06-05 00:00:00  861eff4711a542e4b93843c6dd7febb0   
1           2018-02-06 00:00:00  290c77bc529b7ac935b93aa66c333dc3   
2           2018-06-13 00:00:00  060e732b5b29e8181a18229c7b0b2b5e   
3           2018-04-10 00:00:00  259dac757896d24d7702b9acbbff3f3c   
4           2018-08-15 00:00:00  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix  ... review_comment_title  \
0                     14409  ...                  NaN   
1                      9790  ...                  NaN   
2                      1151  ...                  NaN   
3                      8775  ...                  NaN   
4                     13056  ...        a melhor nota   

                          review_comment_message  review_creation_date  \
0                                            NaN   2017-05-26 00:00:00   
1                                            NaN   2018-01-30 00:00:00   
2                                            NaN   2018-06-15 00:00:00   
3                                            NaN   2018-03-29 00:00:00   
4  O baratheon è esxelente Amo adoro o baratheon   2018-08-10 00:00:00   

  review_answer_timestamp  order_item_id                        product_id  \
0     2017-05-30 22:34:40            1.0  a9516a079e37a9c9c36b9b78b10169e8   
1     2018-02-10 22:43:29            1.0  4aa6014eceb682077f9dc4bffebc05b0   
2     2018-06-15 12:10:59            1.0  bd07b66896d6f1494f5b86251848ced7   
3     2018-04-02 18:36:47            1.0  a5647c44af977b148e0a3a4751a09e2e   
4     2018-08-17 01:59:52            1.0  9391a573abe00141c56e38d84d7d5b3b   

                          seller_id  shipping_limit_date   price freight_value  
0  7c67e1448b00f6e969d365cea6b010ab  2017-05-22 15:22:12  124.99         21.88  
1  b8bc237ba3788b23da09c0f1f3a3288c  2018-01-18 20:58:32  289.00         46.48  
2  7c67e1448b00f6e969d365cea6b010ab  2018-06-05 16:19:10  139.94         17.79  
3  7c67e1448b00f6e969d365cea6b010ab  2018-03-27 16:31:16  149.94         23.36  
4  4a3ca9315b744ce9f8e9374361493884  2018-07-31 10:10:09  230.00         22.25  

[5 rows x 28 columns]

In [24]:
#5. Order Items + Products
orders_products = pd.merge(orders_items, product, on = "product_id", how = "left")
orders_products.head()

order_id                       customer_id  \
0  00e7ee1b050b8499577073aeb2a297a1  06b8999e2fba1a1fbc88172c00ba8bc7   
1  29150127e6685892b6eab3eec79f59c7  18955e83d337fd6b2def6b18a428ac77   
2  b2059ed67ce144a36e2aa97d2c9e9ad2  4e7b3e00288586ebd08712fdd0374a03   
3  951670f92359f4fe4a63112aa7306eba  b2b6027bc5c5109e529d4dc6358b12c3   
4  6b7d50bd145f6fc7f33cebabd7e49d0f  4f2d8ab171c80ec8364f7c12e35b23ad   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-05-16 15:05:35  2017-05-16 15:22:12   
1    delivered      2018-01-12 20:48:24  2018-01-12 20:58:32   
2    delivered      2018-05-19 16:07:45  2018-05-20 16:19:10   
3    delivered      2018-03-13 16:06:38  2018-03-13 17:29:19   
4    delivered      2018-07-29 09:51:30  2018-07-29 10:10:09   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-05-23 10:47:57           2017-05-25 10:35:35   
1          2018-01-15 17:14:59           2018-01-29 12:41:19   
2          2018-06-11 14:31:00           2018-06-14 17:58:51   
3          2018-03-27 23:22:42           2018-03-28 16:04:25   
4          2018-07-30 15:16:00           2018-08-09 20:55:48   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-06-05 00:00:00  861eff4711a542e4b93843c6dd7febb0   
1           2018-02-06 00:00:00  290c77bc529b7ac935b93aa66c333dc3   
2           2018-06-13 00:00:00  060e732b5b29e8181a18229c7b0b2b5e   
3           2018-04-10 00:00:00  259dac757896d24d7702b9acbbff3f3c   
4           2018-08-15 00:00:00  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix  ...   price freight_value  product_category_name  \
0                     14409  ...  124.99         21.88      moveis_escritorio   
1                      9790  ...  289.00         46.48  utilidades_domesticas   
2                      1151  ...  139.94         17.79      moveis_escritorio   
3                      8775  ...  149.94         23.36      moveis_escritorio   
4                     13056  ...  230.00         22.25          casa_conforto   

  product_name_lenght  product_description_lenght  product_photos_qty  \
0                41.0                      1141.0                 1.0   
1                43.0                      1002.0                 3.0   
2                55.0                       955.0                 1.0   
3                48.0                      1066.0                 1.0   
4                61.0                       407.0                 1.0   

  product_weight_g  product_length_cm product_height_cm product_width_cm  
0           8683.0               54.0              64.0             31.0  
1          10150.0               89.0              15.0             40.0  
2           8267.0               52.0              52.0             17.0  
3          12160.0               56.0              51.0             28.0  
4           5200.0               45.0              15.0             35.0  

[5 rows x 36 columns]

In [25]:
#6. Products + Product Category Translation
orders_products = pd.merge(orders_products, product_category, on = "product_category_name", how = "left")
orders_products.head()

order_id                       customer_id  \
0  00e7ee1b050b8499577073aeb2a297a1  06b8999e2fba1a1fbc88172c00ba8bc7   
1  29150127e6685892b6eab3eec79f59c7  18955e83d337fd6b2def6b18a428ac77   
2  b2059ed67ce144a36e2aa97d2c9e9ad2  4e7b3e00288586ebd08712fdd0374a03   
3  951670f92359f4fe4a63112aa7306eba  b2b6027bc5c5109e529d4dc6358b12c3   
4  6b7d50bd145f6fc7f33cebabd7e49d0f  4f2d8ab171c80ec8364f7c12e35b23ad   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-05-16 15:05:35  2017-05-16 15:22:12   
1    delivered      2018-01-12 20:48:24  2018-01-12 20:58:32   
2    delivered      2018-05-19 16:07:45  2018-05-20 16:19:10   
3    delivered      2018-03-13 16:06:38  2018-03-13 17:29:19   
4    delivered      2018-07-29 09:51:30  2018-07-29 10:10:09   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-05-23 10:47:57           2017-05-25 10:35:35   
1          2018-01-15 17:14:59           2018-01-29 12:41:19   
2          2018-06-11 14:31:00           2018-06-14 17:58:51   
3          2018-03-27 23:22:42           2018-03-28 16:04:25   
4          2018-07-30 15:16:00           2018-08-09 20:55:48   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-06-05 00:00:00  861eff4711a542e4b93843c6dd7febb0   
1           2018-02-06 00:00:00  290c77bc529b7ac935b93aa66c333dc3   
2           2018-06-13 00:00:00  060e732b5b29e8181a18229c7b0b2b5e   
3           2018-04-10 00:00:00  259dac757896d24d7702b9acbbff3f3c   
4           2018-08-15 00:00:00  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix  ... freight_value  product_category_name  \
0                     14409  ...         21.88      moveis_escritorio   
1                      9790  ...         46.48  utilidades_domesticas   
2                      1151  ...         17.79      moveis_escritorio   
3                      8775  ...         23.36      moveis_escritorio   
4                     13056  ...         22.25          casa_conforto   

   product_name_lenght product_description_lenght  product_photos_qty  \
0                 41.0                     1141.0                 1.0   
1                 43.0                     1002.0                 3.0   
2                 55.0                      955.0                 1.0   
3                 48.0                     1066.0                 1.0   
4                 61.0                      407.0                 1.0   

   product_weight_g product_length_cm  product_height_cm product_width_cm  \
0            8683.0              54.0               64.0             31.0   
1           10150.0              89.0               15.0             40.0   
2            8267.0              52.0               52.0             17.0   
3           12160.0              56.0               51.0             28.0   
4            5200.0              45.0               15.0             35.0   

  product_category_name_english  
0              office_furniture  
1                    housewares  
2              office_furniture  
3              office_furniture  
4                  home_confort  

[5 rows x 37 columns]

In [26]:
#7. Order Items + Sellers
orders_sellers = pd.merge(orders_products, sellers, on = "seller_id", how = "left")
orders_sellers.head()

order_id                       customer_id  \
0  00e7ee1b050b8499577073aeb2a297a1  06b8999e2fba1a1fbc88172c00ba8bc7   
1  29150127e6685892b6eab3eec79f59c7  18955e83d337fd6b2def6b18a428ac77   
2  b2059ed67ce144a36e2aa97d2c9e9ad2  4e7b3e00288586ebd08712fdd0374a03   
3  951670f92359f4fe4a63112aa7306eba  b2b6027bc5c5109e529d4dc6358b12c3   
4  6b7d50bd145f6fc7f33cebabd7e49d0f  4f2d8ab171c80ec8364f7c12e35b23ad   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-05-16 15:05:35  2017-05-16 15:22:12   
1    delivered      2018-01-12 20:48:24  2018-01-12 20:58:32   
2    delivered      2018-05-19 16:07:45  2018-05-20 16:19:10   
3    delivered      2018-03-13 16:06:38  2018-03-13 17:29:19   
4    delivered      2018-07-29 09:51:30  2018-07-29 10:10:09   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-05-23 10:47:57           2017-05-25 10:35:35   
1          2018-01-15 17:14:59           2018-01-29 12:41:19   
2          2018-06-11 14:31:00           2018-06-14 17:58:51   
3          2018-03-27 23:22:42           2018-03-28 16:04:25   
4          2018-07-30 15:16:00           2018-08-09 20:55:48   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-06-05 00:00:00  861eff4711a542e4b93843c6dd7febb0   
1           2018-02-06 00:00:00  290c77bc529b7ac935b93aa66c333dc3   
2           2018-06-13 00:00:00  060e732b5b29e8181a18229c7b0b2b5e   
3           2018-04-10 00:00:00  259dac757896d24d7702b9acbbff3f3c   
4           2018-08-15 00:00:00  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix  ... product_description_lenght  \
0                     14409  ...                     1141.0   
1                      9790  ...                     1002.0   
2                      1151  ...                      955.0   
3                      8775  ...                     1066.0   
4                     13056  ...                      407.0   

  product_photos_qty  product_weight_g product_length_cm  product_height_cm  \
0                1.0            8683.0              54.0               64.0   
1                3.0           10150.0              89.0               15.0   
2                1.0            8267.0              52.0               52.0   
3                1.0           12160.0              56.0               51.0   
4                1.0            5200.0              45.0               15.0   

   product_width_cm product_category_name_english  seller_zip_code_prefix  \
0              31.0              office_furniture                  8577.0   
1              40.0                    housewares                 88303.0   
2              17.0              office_furniture                  8577.0   
3              28.0              office_furniture                  8577.0   
4              35.0                  home_confort                 14940.0   

       seller_city seller_state  
0  itaquaquecetuba           SP  
1           itajai           SC  
2  itaquaquecetuba           SP  
3  itaquaquecetuba           SP  
4         ibitinga           SP  

[5 rows x 40 columns]

In [28]:
#8. Customers + Geolocation (via zip prefix)
# Note: many geolocations per ZIP, so we take first occurence per prefix

geo_unique = geolocation.groupby("geolocation_zip_code_prefix").first().reset_index()

final_df = pd.merge(
    orders_sellers,
    geo_unique,
    left_on = "customer_zip_code_prefix",
    right_on = "geolocation_zip_code_prefix",
    how = "left"
)

print(final_df.shape)
final_df.head()

(119143, 45)


order_id                       customer_id  \
0  00e7ee1b050b8499577073aeb2a297a1  06b8999e2fba1a1fbc88172c00ba8bc7   
1  29150127e6685892b6eab3eec79f59c7  18955e83d337fd6b2def6b18a428ac77   
2  b2059ed67ce144a36e2aa97d2c9e9ad2  4e7b3e00288586ebd08712fdd0374a03   
3  951670f92359f4fe4a63112aa7306eba  b2b6027bc5c5109e529d4dc6358b12c3   
4  6b7d50bd145f6fc7f33cebabd7e49d0f  4f2d8ab171c80ec8364f7c12e35b23ad   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-05-16 15:05:35  2017-05-16 15:22:12   
1    delivered      2018-01-12 20:48:24  2018-01-12 20:58:32   
2    delivered      2018-05-19 16:07:45  2018-05-20 16:19:10   
3    delivered      2018-03-13 16:06:38  2018-03-13 17:29:19   
4    delivered      2018-07-29 09:51:30  2018-07-29 10:10:09   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-05-23 10:47:57           2017-05-25 10:35:35   
1          2018-01-15 17:14:59           2018-01-29 12:41:19   
2          2018-06-11 14:31:00           2018-06-14 17:58:51   
3          2018-03-27 23:22:42           2018-03-28 16:04:25   
4          2018-07-30 15:16:00           2018-08-09 20:55:48   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-06-05 00:00:00  861eff4711a542e4b93843c6dd7febb0   
1           2018-02-06 00:00:00  290c77bc529b7ac935b93aa66c333dc3   
2           2018-06-13 00:00:00  060e732b5b29e8181a18229c7b0b2b5e   
3           2018-04-10 00:00:00  259dac757896d24d7702b9acbbff3f3c   
4           2018-08-15 00:00:00  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix  ... product_width_cm  \
0                     14409  ...             31.0   
1                      9790  ...             40.0   
2                      1151  ...             17.0   
3                      8775  ...             28.0   
4                     13056  ...             35.0   

  product_category_name_english  seller_zip_code_prefix      seller_city  \
0              office_furniture                  8577.0  itaquaquecetuba   
1                    housewares                 88303.0           itajai   
2              office_furniture                  8577.0  itaquaquecetuba   
3              office_furniture                  8577.0  itaquaquecetuba   
4                  home_confort                 14940.0         ibitinga   

   seller_state  geolocation_zip_code_prefix geolocation_lat  geolocation_lng  \
0            SP                      14409.0      -20.509897       -47.397866   
1            SC                       9790.0      -23.726853       -46.545746   
2            SP                       1151.0      -23.527788       -46.660310   
3            SP                       8775.0      -23.496930       -46.185352   
4            SP                      13056.0      -22.987222       -47.151073   

        geolocation_city geolocation_state  
0                 franca                SP  
1  sao bernardo do campo                SP  
2              sao paulo                SP  
3        mogi das cruzes                SP  
4               campinas                SP  

[5 rows x 45 columns]

# **PROBABILITY**
<hr>
<hr>

1.  Probability of an Event
<br>
Concept: Basic probability = favorable outcomes / total outcomes
<br>
Question: What is the probability that a randomly chosen order received a 5-star review?

In [29]:
# we will be using the order_review data to answer this question

orders_reviews.head(2)

order_id                       customer_id  \
0  00e7ee1b050b8499577073aeb2a297a1  06b8999e2fba1a1fbc88172c00ba8bc7   
1  29150127e6685892b6eab3eec79f59c7  18955e83d337fd6b2def6b18a428ac77   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-05-16 15:05:35  2017-05-16 15:22:12   
1    delivered      2018-01-12 20:48:24  2018-01-12 20:58:32   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-05-23 10:47:57           2017-05-25 10:35:35   
1          2018-01-15 17:14:59           2018-01-29 12:41:19   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-06-05 00:00:00  861eff4711a542e4b93843c6dd7febb0   
1           2018-02-06 00:00:00  290c77bc529b7ac935b93aa66c333dc3   

   customer_zip_code_prefix  ... payment_sequential payment_type  \
0                     14409  ...                1.0  credit_card   
1                      9790  ...                1.0  credit_card   

   payment_installments payment_value                         review_id  \
0                   2.0        146.87  88b8b52d46df026a9d1ad2136a59b30b   
1                   8.0        335.48  02fc48a9efa3e3d0f1a8ea26507eeec3   

   review_score review_comment_title  review_comment_message  \
0           4.0                  NaN                     NaN   
1           5.0                  NaN                     NaN   

  review_creation_date review_answer_timestamp  
0  2017-05-26 00:00:00     2017-05-30 22:34:40  
1  2018-01-30 00:00:00     2018-02-10 22:43:29  

[2 rows x 22 columns]

In [34]:
# Probability of a 5-star review
# out of all orders, what is the probability that a review has a 5-star rating?
# this tells us how likely a customer is to give the maximum statisfaction score.

# p(5star) = number of 5-stars review/total reviews

total_reviews = len(order_reviews)

favourable_outcomes = (order_reviews["review_score"] == 5).sum()

p_5star = favourable_outcomes / total_reviews

print(f"Probability of 5 stars = {round(p_5star *100, 2)}%")

Probability of 5 stars = 57.78%


**Interpretation of result** <br>
Approximately 58% of customers give 5 stars. This is the baseline "success" probability of excellent reviews

<hr>
<hr>

2.  Complementary Probability <br>
Concept: _`P(A complement) = 1-P(A)`_ <br>
Question: What is the probability that a review is not 5 stars?

In [36]:
# if we know the probability of 5 stars, what's the probability of not getting a 5-star review?
# this means any rating from 1 to 4.

# we are still going to use the review dataset
# formula style
p_not5 = (orders_reviews["review_score"] != 5).sum() / len(order_reviews)

print(f"P(Review != 5 stars) = {round(p_not5 * 100, 2)}")

P(Review != 5 stars) = 45.01


**Interpretation** <br>
Approximately 42% of reviews are less than perfect. This highlights that not everyone is fully satisfied with product purchased

<hr>
<hr>

3.  Joint Probability
Concept: _P(A∩B)_ <br>
Question: What is the probability that an order was paind by credit card and received a 5-star review?


In [37]:
# this gives the likelihood of two events happening together

# here we will be using the orders_payments and orders_reviews. And we have merged them already
# lets merge orders_payments and order_reviews

order_reviews.head(2)

orders_payment_reviews = pd.merge(orders_reviews, orders_payments, on = "order_id", how = "left")
orders_payment_reviews.head()

order_id                     customer_id_x  \
0  00e7ee1b050b8499577073aeb2a297a1  06b8999e2fba1a1fbc88172c00ba8bc7   
1  29150127e6685892b6eab3eec79f59c7  18955e83d337fd6b2def6b18a428ac77   
2  b2059ed67ce144a36e2aa97d2c9e9ad2  4e7b3e00288586ebd08712fdd0374a03   
3  951670f92359f4fe4a63112aa7306eba  b2b6027bc5c5109e529d4dc6358b12c3   
4  6b7d50bd145f6fc7f33cebabd7e49d0f  4f2d8ab171c80ec8364f7c12e35b23ad   

  order_status_x order_purchase_timestamp_x  order_approved_at_x  \
0      delivered        2017-05-16 15:05:35  2017-05-16 15:22:12   
1      delivered        2018-01-12 20:48:24  2018-01-12 20:58:32   
2      delivered        2018-05-19 16:07:45  2018-05-20 16:19:10   
3      delivered        2018-03-13 16:06:38  2018-03-13 17:29:19   
4      delivered        2018-07-29 09:51:30  2018-07-29 10:10:09   

  order_delivered_carrier_date_x order_delivered_customer_date_x  \
0            2017-05-23 10:47:57             2017-05-25 10:35:35   
1            2018-01-15 17:14:59             2018-01-29 12:41:19   
2            2018-06-11 14:31:00             2018-06-14 17:58:51   
3            2018-03-27 23:22:42             2018-03-28 16:04:25   
4            2018-07-30 15:16:00             2018-08-09 20:55:48   

  order_estimated_delivery_date_x              customer_unique_id_x  \
0             2017-06-05 00:00:00  861eff4711a542e4b93843c6dd7febb0   
1             2018-02-06 00:00:00  290c77bc529b7ac935b93aa66c333dc3   
2             2018-06-13 00:00:00  060e732b5b29e8181a18229c7b0b2b5e   
3             2018-04-10 00:00:00  259dac757896d24d7702b9acbbff3f3c   
4             2018-08-15 00:00:00  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix_x  ... order_delivered_customer_date_y  \
0                       14409  ...             2017-05-25 10:35:35   
1                        9790  ...             2018-01-29 12:41:19   
2                        1151  ...             2018-06-14 17:58:51   
3                        8775  ...             2018-03-28 16:04:25   
4                       13056  ...             2018-08-09 20:55:48   

  order_estimated_delivery_date_y              customer_unique_id_y  \
0             2017-06-05 00:00:00  861eff4711a542e4b93843c6dd7febb0   
1             2018-02-06 00:00:00  290c77bc529b7ac935b93aa66c333dc3   
2             2018-06-13 00:00:00  060e732b5b29e8181a18229c7b0b2b5e   
3             2018-04-10 00:00:00  259dac757896d24d7702b9acbbff3f3c   
4             2018-08-15 00:00:00  345ecd01c38d18a9036ed96c73b8d066   

  customer_zip_code_prefix_y        customer_city_y  customer_state_y  \
0                      14409                 franca                SP   
1                       9790  sao bernardo do campo                SP   
2                       1151              sao paulo                SP   
3                       8775        mogi das cruzes                SP   
4                      13056               campinas                SP   

  payment_sequential_y  payment_type_y payment_installments_y payment_value_y  
0                  1.0     credit_card                    2.0          146.87  
1                  1.0     credit_card                    8.0          335.48  
2                  1.0     credit_card                    7.0          157.73  
3                  1.0     credit_card                    1.0          173.30  
4                  1.0     credit_card                    8.0          252.25  

[5 rows x 37 columns]

In [38]:
orders_payment_reviews.columns

Index(['order_id', 'customer_id_x', 'order_status_x',
       'order_purchase_timestamp_x', 'order_approved_at_x',
       'order_delivered_carrier_date_x', 'order_delivered_customer_date_x',
       'order_estimated_delivery_date_x', 'customer_unique_id_x',
       'customer_zip_code_prefix_x', 'customer_city_x', 'customer_state_x',
       'payment_sequential_x', 'payment_type_x', 'payment_installments_x',
       'payment_value_x', 'review_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp', 'customer_id_y', 'order_status_y',
       'order_purchase_timestamp_y', 'order_approved_at_y',
       'order_delivered_carrier_date_y', 'order_delivered_customer_date_y',
       'order_estimated_delivery_date_y', 'customer_unique_id_y',
       'customer_zip_code_prefix_y', 'customer_city_y', 'customer_state_y',
       'payment_sequential_y', 'payment_type_y', 'payment_installments_y',
       'payment_value_y'],
      dt

In [41]:
total_orders = len(orders_payment_reviews)

favourable_joint = ((orders_payment_reviews["payment_type_y"] == "credit_card") & 
                    (orders_payment_reviews["review_score"] == 5)).sum()

p_joint = favourable_joint / total_orders

print(f"P(Credit card ∩ 5 stars) = {round(p_joint * 100, 2)}%")

P(Credit card ∩ 5 stars) = 37.51%


**Interpretation**<br>
Approximately 38% of all orders are both paid by credit card and get a perfect review.
<hr>
<hr>

4.  Conditional Probability<br>
Concept: <br>
_`P(A|B) = P(A∩B) / P(B)`_ <br>
_`P(A|B) = favourable_outcomes / total_credit_card_outcomes`_  <br>
Question: What is the probability that a review is 5 stars given that the payment was made by credit_card?

In [42]:
# if we already know the payment method was credit card, what is the probability?
# this measures satisfaction among credit card users only.
# we are solving for independent conditional probability

# we might took few steps, and solve beond the question for proper understanding
# that is, P(5 stars | payment method = X)

#P(5|credict_card)
#P(5|voucher)
#P(5|boleto)

# each is a separate conditional distribution. They are not mutually exclusive outcomes of the same experiment.

favourable_cond = ((orders_payment_reviews["payment_type_y"] == "credit+card") &
                   (orders_payment_reviews["review_score"] == 5)).sum()

total_credit = (orders_payment_reviews["payment_type_y"] == "credit_card").sum()

p_cond = favourable_cond / total_credit

print(f"P(5 stars | Credit Card) = {round(p_cond * 100, 2)}%")

# output: P(5 stars | Credit Card) = 57.39%

favourable_cond = ((orders_payment_reviews["payment_type_y"] == "voucher") &
                   (orders_payment_reviews["review_score"] == 5)).sum()

total_credit = (orders_payment_reviews["payment_type_y"] == "voucher").sum()

p_cond = favourable_cond / total_credit

print(f"P(5 stars | voucher) = {round(p_cond * 100, 2)}%")

# output: P(5 stars | voucher) = 55.71%

favourable_cond = ((orders_payment_reviews["payment_type_y"] == "boleto") &
                   (orders_payment_reviews["review_score"] == 5)).sum()

total_credit = (orders_payment_reviews["payment_type_y"] == "boleto").sum()

p_cond = favourable_cond / total_credit
print(f"P(5 stars | boleto) = {round(p_cond * 100, 2)}%")

# output: P(5 stars | boleto) = 56.95%

P(5 stars | Credit Card) = 0.0%
P(5 stars | voucher) = 55.71%
P(5 stars | boleto) = 56.95%


**Interpretation** <br>
Each one is around 55-57%, meaning in every payment method, about half of cusomers gave 5 stars.

<hr>
<hr>

6.  Bayes' Rule<br>
Concept:<br>
_`P(A|B) = P(B|A).P(A) / P(B)`_ <br>
A = payment <br>
B = review = 5 <br>

_`P(payment | review = 5) = P(review = 5 | payment).p(payment) / p(review = 5)`_<br>
Question: Given that an order received a 5-star review, what is the probability that it was paid with each payment method (credit_card, boleto, voucher)? <br>
Among only the 5-star reviews, how are payment methods distributed?

In [43]:
# among orders that got a 5-star review, what fraction used each payment method (credit_card, boleto, voucher, etc.)?
# that is, if we look at only 5-star reviews, how are the split across payment types?

# total number of 5-star reviews (denominator for all)
total_5star = (orders_payment_reviews["review_score"] == 5).sum()

#-- P(payment = credit card | review = 5) ----
favourable_cond = ((orders_payment_reviews["payment_type_y"] == "credit_card") & 
                   (orders_payment_reviews["review_score"] == 5)).sum()

p_cond = favourable_cond / total_5star
print(f"P(Credit Card | 5 stars) = {round(p_cond * 100, 2)}%")

#-- P(payment = voucher | review = 5)---
favourable_cond = ((orders_payment_reviews["payment_type_y"] == "voucher") & 
                   (orders_payment_reviews["review_score"] == 5)).sum()

p_cond = favourable_cond / total_5star
print(f"P(Voucher | 5 stars) = {round(p_cond * 100, 2)}%")

#-- P(payment = boleto | review = 5) ---
favourable_cond = ((orders_payment_reviews["payment_type_y"] == "boleto") &
                   (orders_payment_reviews["review_score"] == 5)).sum()

p_cond = favourable_cond / total_5star
print(f"P(Boleto | 5 stars) = {round(p_cond * 100, 2)}%")

P(Credit Card | 5 stars) = 65.74%
P(Voucher | 5 stars) = 16.9%
P(Boleto | 5 stars) = 16.05%


**Interpretation**
1.  `P(Credit Card | 5 stars)` = Among all 5-star reviews, what fraction used credit card? Among 5-star reviews, 65.74$ used credit card
2.  `P(Voucher | 5 stars)` = Among all 5-star reviews, what fraction used voucher? Among 5-star reviews, `16.9%` used boleto.
3.  `P(Boleto | 5 stars)` = Among all 5-star reviews, what fraction used boleto? `16.05%` of 5-star reviews uses Boleto payment method.

<br>

Unlike `P(5 | payment)` - conditional probability, these will sum to approximately 100% across all payment methods, because every 5-star order must fall into exactly one payment category <br>
<br>
**I hope the interpretation or difference is clear enough, if not, please go through it again**

### **Expectation and Variance**
**Expectation**
-   This is what you would expect from an experiment if you repeat an experiment many times. In practice, this means taking the average of the outcome of an experiment, the resulting value is the expectation. In statistics speak, we are talkinb about the mean (the centre of the distribution).
-   So we talk about the baseline performance, they are actually talking about expectation.
-   Examples of performance are average delivery time, average customer spend, average rating
-   We use it to find out;
    *   How long a customer typically wait to get his or her delivery?
    *   What is the expected revenue per order?
    *   What is the overall satisfaction level
    *   What is the expected sale per seller?


**Variance**
-   This explain how far the result of an experiment is from what is expected. When this value is very high, we would conclude that outcome of the experiement is not consistent, it is not predictable, it is unstable. When the value is low, it speaks of predictability, consistency, and stability. In summary, the wider the spread the higher the risk, the narrower the spread the lower the risk and reliability.
-   Note that variace is more interpretable when it is converted to standard deviation (same unit as your data)

**Usecase**
-   The combination of these two are use for risk assessment
-   They are also used for measuring reliability - how consistent are deliveries, ragings, or payments?
-   They are also use to make data driven decistion - To choose sellers, couriers, products with stable performance not just high averages.
-   Uncertainty estimation (used in confidence intervals and hypothesis testing)
-   Machine learning (feature scaling and PCA rely on mean & variance)
-   **Let's see some examples**
    *   Mean delivery time = 10 days, SD = 2 days
        -   Meaning that "Most customers get deliveries between 8-12 days. Service is predictable."
    *   Mean delivery time = 10 days, SD = 6 days
        -   Meaning that "Some customers get deliveries in 4 days, others in 16. Unreliable service."
    *   Mean review score = 4.2, SD = 0.5
        -   "Almost everyone gives between 4-5 stars. Strong and consistent reputation."
    *   Mean review score = 4.2, SD = 1.8
        -   "Some give 5, others 1. The average hides mixed experiences."

**Question 1**<br>
What is the average delivery time (expectation) and variance (spread/consistency) of deliveries

In [44]:
orders_payment_reviews.head()

order_id                     customer_id_x  \
0  00e7ee1b050b8499577073aeb2a297a1  06b8999e2fba1a1fbc88172c00ba8bc7   
1  29150127e6685892b6eab3eec79f59c7  18955e83d337fd6b2def6b18a428ac77   
2  b2059ed67ce144a36e2aa97d2c9e9ad2  4e7b3e00288586ebd08712fdd0374a03   
3  951670f92359f4fe4a63112aa7306eba  b2b6027bc5c5109e529d4dc6358b12c3   
4  6b7d50bd145f6fc7f33cebabd7e49d0f  4f2d8ab171c80ec8364f7c12e35b23ad   

  order_status_x order_purchase_timestamp_x  order_approved_at_x  \
0      delivered        2017-05-16 15:05:35  2017-05-16 15:22:12   
1      delivered        2018-01-12 20:48:24  2018-01-12 20:58:32   
2      delivered        2018-05-19 16:07:45  2018-05-20 16:19:10   
3      delivered        2018-03-13 16:06:38  2018-03-13 17:29:19   
4      delivered        2018-07-29 09:51:30  2018-07-29 10:10:09   

  order_delivered_carrier_date_x order_delivered_customer_date_x  \
0            2017-05-23 10:47:57             2017-05-25 10:35:35   
1            2018-01-15 17:14:59             2018-01-29 12:41:19   
2            2018-06-11 14:31:00             2018-06-14 17:58:51   
3            2018-03-27 23:22:42             2018-03-28 16:04:25   
4            2018-07-30 15:16:00             2018-08-09 20:55:48   

  order_estimated_delivery_date_x              customer_unique_id_x  \
0             2017-06-05 00:00:00  861eff4711a542e4b93843c6dd7febb0   
1             2018-02-06 00:00:00  290c77bc529b7ac935b93aa66c333dc3   
2             2018-06-13 00:00:00  060e732b5b29e8181a18229c7b0b2b5e   
3             2018-04-10 00:00:00  259dac757896d24d7702b9acbbff3f3c   
4             2018-08-15 00:00:00  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix_x  ... order_delivered_customer_date_y  \
0                       14409  ...             2017-05-25 10:35:35   
1                        9790  ...             2018-01-29 12:41:19   
2                        1151  ...             2018-06-14 17:58:51   
3                        8775  ...             2018-03-28 16:04:25   
4                       13056  ...             2018-08-09 20:55:48   

  order_estimated_delivery_date_y              customer_unique_id_y  \
0             2017-06-05 00:00:00  861eff4711a542e4b93843c6dd7febb0   
1             2018-02-06 00:00:00  290c77bc529b7ac935b93aa66c333dc3   
2             2018-06-13 00:00:00  060e732b5b29e8181a18229c7b0b2b5e   
3             2018-04-10 00:00:00  259dac757896d24d7702b9acbbff3f3c   
4             2018-08-15 00:00:00  345ecd01c38d18a9036ed96c73b8d066   

  customer_zip_code_prefix_y        customer_city_y  customer_state_y  \
0                      14409                 franca                SP   
1                       9790  sao bernardo do campo                SP   
2                       1151              sao paulo                SP   
3                       8775        mogi das cruzes                SP   
4                      13056               campinas                SP   

  payment_sequential_y  payment_type_y payment_installments_y payment_value_y  
0                  1.0     credit_card                    2.0          146.87  
1                  1.0     credit_card                    8.0          335.48  
2                  1.0     credit_card                    7.0          157.73  
3                  1.0     credit_card                    1.0          173.30  
4                  1.0     credit_card                    8.0          252.25  

[5 rows x 37 columns]

In [45]:
# first convert the datatime columns to pandas datetime

# convert to datetime
orders_payment_reviews["order_purchase_timestamp_x"] = pd.to_datetime(orders_payment_reviews["order_purchase_timestamp_x"])
orders_payment_reviews["order_delivered_customer_date_x"] = pd.to_datetime(orders_payment_reviews["order_delivered_customer_date_x"])

# calculate delivery time in days
orders_payment_reviews["delivery_time_days"] = (orders_payment_reviews["order_delivered_customer_date_x"] - orders_payment_reviews["order_purchase_timestamp_x"]).dt.days

# drop missing values
delivery_times = orders_payment_reviews["delivery_time_days"].dropna()

# lets get the expectation (mean)
mean_delivery = delivery_times.mean()

# variance & standard deviation
variance_delivery = delivery_times.var()
std_delivery = delivery_times.std()

print("Expected delivery time (Mean):", round(mean_delivery, 2), "days")
print("Variance of delivery time:", round(variance_delivery, 2))
print("Standard deviation of delivery time:", round(std_delivery, 2), "days")

Expected delivery time (Mean): 12.13 days
Variance of delivery time: 89.28
Standard deviation of delivery time: 9.45 days


**Interpretation**

**Question 2**

What is the expected customer review score and how much do reviews vary?

In [ ]:
# expectation (mean review score)
mean_review = orders_payment_reviews["review_score"].mean()

# variance & standard deviation
variance_review = orders_payment_reviews["review_score"].var()
std_review = orders_payment_reviews["review_score"].std()

print("Expected review score (mean):", round(mean_review, 2))
print("Variance of review score:", round(variance_review, 2))
print("Standard deviation of review score:", round(std_review), 2)

**Interpretation**

The mean review score shows the overall satisfaction (e.g., ~4 stars = mostly happy customers). The std tells us that most customer give between 3 and 5 stars. If STD was very high (e.g., 2), it would mean customers are divided (some give 1, others give 5). but here, the service reputation is fairly consistent

### **Sampling & Estimation**

Random sampling means selecting a subset of your data without bias so that it fairly represents the whole population.

In business terms, imagine you can't survey all customers. Instead, you ask a random group and use their responses to estimate the overall trend.
-   Let's see how we can work with this in our dataset

In [ ]:
orders_payment_reviews.head(2)

In [ ]:
# population mean (true mean from entire dataset)
pop_mean_payment = orders_payment_reviews["payment_value_y"].mean()

# lets take a random sample of 50 orders
sample_50 = orders_payment_reviews["payment_value_y"].sample(50, random_state = 42)
sample_mean_50 = sample_50.mean()

# take a random sample of 200 orders
sample_200 = orders_payment_reviews["payment_value_y"].sample(200, random_state = 42)
sample_mean_200 = sample_200.mean()

print("Population mean payment value:", round(pop_mean_payment, 2))
print("Sample mean (n=50):", round(sample_mean_50, 2))
print("Sample mean (n=200):", round(sample_mean_200, 2))

**Interpreation**
*   Population mean = "the real average spending across all customers."
*   Sample mean (50 orders) may be off (underestimate or overestimate).
*   Sample mean (200 orders) should be closer to the population mean.

This shows that why random sampling is powerful we can get a good estimate of the population without looking at every order.

### Law of Large Numbers (LLN)

You might wonder why big data is a reliable estimate, the idea behind it is the law of large numbers. Imagine you are asking people are the nature of the AI fellowship, the more people you ask, the more likely to build up your expectation or mindset about what the quality of the fellowship.

*   The Law of Large Numbers says,

"As the sample size increases, the sample mean --> converges to the true population mean"

*   Let us experiment a bit...

In [ ]:
import numpy as np

# draw increasing sample sizes and track their means
sample_sizes = [10, 50, 100, 500, 1000, 5000]
sample_means = []

for n in sample_sizes:
    sample = orders_payment_reviews["payment_value_y"].sample(n, random_state = 234)
    sample_means.append(sample.mean())

# show results
for n, mean in zip(sample_sizes, sample_means):
    print(f"Sample size = {n}, Sample mean = {round(mean, 2)}")

print("Populationi mean (true):", round(pop_mean_payment, 2))

**Interpretation**

The result above should explain the reliability of big data, right?

With 10 samples, the mean may jump around (unreliable). Ad with 1000+ samples the mean is almost the same as population mean.

### **Central Limit Theorem (CLT)**

In [ ]:
orders_payment_reviews.head(2)

In [ ]:
orders_payment_reviews.columns

In [ ]:
# lets collect 100 samples of review_scores, take the mean. Repeat the same process a thousand times.
# then take all the 1000 means use it to plot a histogram and lets see what the shape or distribution would look like.

# collect many sample means
sample_means = []
for i in range(1000):   # repeat 1000 times
    sample = orders_payment_reviews["review_score"].sample(100, replace = True)
    sample_means.append(sample.mean())

# plot the sampling distribution of the mean
plt.hist(sample_means, bins = 30, edgecolor = "black")
plt.axvline(np.mean(sample_means), color = "red", linestyle = "dashed", linewidth = 2)
plt.title("Central Limit Theorem: Sampling Distribution of Review Score Means (n = 100)")
plt.xlabel("Sample mean of review_score")
plt.ylabel("Frequency")
plt.show()

**The Central Limit Theorem says**, _"If we take many random samples and calculate their means the distribution of those means will look like a normal distribution (bell curve), even if the original data is not normal."_

It allows us to use normal distribution tools (confident intervals, hypothesis testing) in Data Science.

Now from out plot, even though review_score is discrete (1-5 stars) and not perfectly normal the distribution of sample means (averages of samples) will look like a bell curve.

This explains why in Data Science we confidently apply normal-based methods (t-tests, regression assumptions, confidence intervals).

### **One-Sample Test**

**A hypothesis** is just an assumption or claim through a population (the "big picture data") that we want to test using a sample (the data we actually have).
-   Let think of it like this;
-   It is like asking, "does this medicine work?"
-   "Do credit card users spend more than boleto users?"
-   "Are customers happier than neutral (score 3)?"

Two be able to answer this question, we have to **"set up two competing statements"**

**The first statement is called Ho (pronouced as H-not) or called the Null Hypothesis**
    -   You can see it as the boring assumption.
    -   It says, "nothing special is happening, things are the same, things are neutral, the assumption is null"
    -   This assumption is usually viewed as TRUE!
    -   Because of this, we really need to conduct a test, this test is called the test of hypothesis.
    -   If the evidence from the test is strong enough, then we have to `reject Ho`
    -   We know that this evidence is not strong enough if the p-value is greater that the experiment score

**The second statement is called H (pronounced as H-one) or called the Alternative Hypothesis**
*   You can see it as the interesting assumption.
*   It argues, "something different is happening, things are not the same, things are not neutral."
*   This is the claim we actually want to investigate.
*   If the evidence from the test is strong enough, then we reject Ho and say H1 is more likely true.

For example:

`If Ho says "average review score = 3 (neutral)"`

`then H1 argues "average review score != 3 (not neutral)."`

*   we will investigate H1 to check.

`If Ho says "credit card spend = boleto spend"`

`then H1 argues "credit card spend != boleto spend."`

So having understand the concept of hypothesis testing, let go back to **A one-sample test.** It is used when we want to check if average (mean) of our sample data is different from some reference value (what we expect, or a "neutral" baseline).

`Ho (Null Hypothesis) - says the sample mean is equal to the reference value`

`H1 (Alternative hypothesis) - says the sample mean is differet from the reference value.`

*   **Lets try setting up an hypothesis now...**

Suppose we want to test if customers are happier than neutral.

In our Olist data, the review_score ranges from 1-5. Neutral is 3. We want to test, "Is the average review score different from 3?"

**Step 1 - Set up hypothesis**

`Ho: Average review score = 3 (customers are neutral)`

`H1: Average review score != 3 (customers are not neutral, they are either happier or unhappier)`

**Step 2 - Why test it?**
-   If Ho is true, Olist has "just okay" service.
-   If H1 is true, Customers are clearly happy (if mean > 3) or unhappy (if mean < 3).

**Question 1**

In the Olist e-commerce dataset, customer reviews are scored from 1 (worst) to 5 (best). A score of 3 is considered neutral. Using the review data, test whether the average customer review score is significantly different from 3 (neutral)

In [ ]:
# we have one sample, that is the set of all customer review scores.

# we are comparing its mean against a benchmark/reference value = 3.

# now, the question is essentially, "Are Olist customers, on average, happier than neutral (3)?"

# lets write the hypothesis

# Ho (Null hypothesis): the average review score is equal to 3
# H1 (alternative hypothesis): the average review score is not equal to 3

# extract review scores
review_scores = order_reviews["review_score"].dropna()

# define the "neutral benchmark"
neutral_score = 3

# perform one-sample t-test
t_stat, p_value = stats.ttest_1samp(review_scores, popmean = neutral_score)

print("Mean review score:", round(review_scores.mean(), 2))
print("T-statistic:", round(t_stat, 3))
print("P-value:", p_value)

# interpretation
if p_value < 0.05:
    print("Reject Ho: The average review score is significantly different from 3 (neutral).")
else:
    print("Fail to reject Ho: The data does not provide enough evidence that the average review differs from 3.")

**Interpretation**

The p-value is ~0, way below 0.05 (strong evidence)

So we conclude that the customers are significantly happier than neutral.

### **Two-Sample t-test**

A two-sample t-test is a statistical test used to compare the average (mean) of two independent groups to see if they are significantly different from each other.

It answers the queestion, "Do two groups have the same average, or are their averages different?"

Null and Alternative Hypothesis (General Form)
Ho (Null Hypothesis) - The two group means are equal.

`𝜇1=𝜇2`​

H1 (Alternative Hypothesis) - the two group means are not equal.

`𝜇1≠𝜇2`


lets illustrate it this way... Two groups - the test always involves two separate groups of data or categories.

Example: Men vs Women, Credit Card vs Boleto, Group A vs Group B.

To compute, we are not looking at individual values, but the overall group average.

But first, the groups have to be unrelated (independent samples) e.g, one customer can't belong to both payment methods at the same time.

How do we interprete the result,

-   If the `p-value < 0.05` means the difference in means is significant.
-   If `p-value >= 0.05` means the difference could just be due to chance.

**Let's have some illustrations here**

Just imagine that your are the manager of Olist e-commerce platform. And you wish to know the customer's payment preference from a data driven approach.

Yea? So you ask the question, "Do customers who pay with credit card spend more money than those who pay with Boleto (Bank slip)?"

-   Group 1 - Customers who use credit card
-   Group 2 - Customers who use Boleto

Variable we care about - payment_value

**Lets define our Hypothesis**
Ho (Null Hypothesis) - On average, Credit card users and Boleto users spend the same amount.

`𝜇(credit) = 𝜇(boleto)`

H1 (Alternative Hypothesis): "On average, Credit Card users spend differently than Boleto users."

`𝜇(credit) ≠ 𝜇(boleto)`

But, wait o, why can't we just compare the averages directly?

The answer is because differences could happen by chance, maybe we sampled just a few expensive orders on Credit Card. So the t-test helps us check if the difference is statistically signinficant (real) or just random noise.

**Question 1**
-   Extract the payment_value for oders paid with Credit Card and Boleto.
    *   Perform a two-sample t-test and interpret the results
        -   What are the average spends for each group?
        -   What is the p-value?
        -   Based on the p-value, do we reject or fail to reject the null hypothesis?
        -   What does this mean in practical terms about customer spending behavior?

In [4]:
# lets craft the hypothesis
# Ho (Null hypothesis) - the average payment_value of credit card orders is equal to the average payment_value of Boleto orders

# H1 (alternative hypothesis) - the average payment_value of credit card orders is different from the average payment_value of Boleto orders.

In [ ]:
# filter payment values by payment type
credit_card = orders_payment_reviews.loc[
    orders_payment_reviews["payment_type_y" == "credit_card", "payment_value_y"]
]

boleto = orders_payment_reviews.loc[
    orders_payment_reviews["payment_type_y"] == "boleto", "payment_value_y"
]

# drop missing values - just in case
credit_card = credit_card.dropna()
boleto = boleto.dropna()

print("Credit Card Average Spend:", round(credit_card.mean(), 2))
print("Boleto Average Spend:", round(boleto.mean(), 2))

In [ ]:
# lets perform our t-test
t_stat, p_value = stats.ttest_ind(credit_card, boleto)

print("T-statistic:", t_stat)
print("P-value:", p_value)

**Interpretation**

**Case 1**

`p-value < 0.05 (Reject Ho)`
-   This means there is a significant difference in spending between Credit Card and Boleto customers.

**Case 2**

`p-value >= 0.05 (Fail to reject Ho)`
-   This means we do not have enough evidence to say the spending is different

In [ ]:
T_statistic = 8.12

p_value = 4.63 * 10 ** -16 # (which is basically)

# let compare the values using p-value with significance level (a = 0.05)

if p_value < 0.05:
    print("Reject Ho: The average payment_value of Credit card orders is different from the average payment_value of Boleto orders.")
else:
    print("Fail to reject Ho: The data does not provide enough evidence that the average payment_value of Boleto orders")

Ho said that Credit Card users spend the same on average as Boleto users.

H1 argued that Credit card users spend differently than Boleto users

From our test, t-test showed that the difference is statistically significant (it is very unlikely to be a random chance).

**Lets take our interpretation further**

Customers who pay with Credit Card spend significantly more on average than those who pay with Boleto.

So this insight can help Olist to;
1.  Prioritize Credit Card paymment promotions, since those customers bring more revenue.
2.  Investigate why Boleto customers spend less, maybe boleto is used for smaller, lower-value purchases.